bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v3/PageImpl.java

```diff
@@ -0,0 +1,451 @@
+            rawPageFilters = pageFilterParameter.getString();
+        }
+
+        if (currentStyle != null && StringUtils.isBlank(rawPageFilters)) {
+            rawPageFilters = currentStyle.get(PN_STRUCTURE_PATTERNS, String.class);
+        }
+
+        if (StringUtils.isBlank(rawPageFilters)) {
+            return Collections.emptyList();
+        }
+
+        String[] pageFilters = rawPageFilters.split(",");
+
+        List<Pattern> pageFilterPatterns = new ArrayList<>();
+        for (String pageFilter : pageFilters) {
+            pageFilterPatterns.add(Pattern.compile(pageFilter));
+        }
+
+        return pageFilterPatterns;
+    }
+
+    /**
+     * Returns a flat list of all the child pages of a given page
+     *
+     * @param page page from which to extract child pages
+     * @param slingRequest request
+     * @param structurePatterns patterns to filter child pages
+     * @param depth depth of the traversal
+     * @return {@link List}
+     */
+    @Nonnull
+    private List<com.day.cq.wcm.api.Page> getChildPageRecursive(com.day.cq.wcm.api.Page page, SlingHttpServletRequest slingRequest, List<Pattern> structurePatterns, int depth) {
+        // By default the value is 0 meaning we do not expose child pages
+        // If the value is set as a positive number it is going to be exposed until the counter is brought down to 0
+        // If the value is set to a negative value all the child pages will be exposed (full traversal tree - aka infinity)
+        // Child pages do not expose their respective child pages
+        if (page == null || depth == 0 || Boolean.TRUE.equals(slingRequest.getAttribute(ATTR_IS_CHILD_PAGE))) {
+            return Collections.emptyList();
+        }
+
+        List<com.day.cq.wcm.api.Page> pages = new ArrayList<>();
+        Iterator<com.day.cq.wcm.api.Page> childPagesIterator = page.listChildren();
+
+        if (childPagesIterator == null || !childPagesIterator.hasNext()) {
+            return Collections.emptyList();
+        }
+
+        // we are about to explore one lower level down the tree
+        depth--;
+
+        boolean noPageFilters = structurePatterns.isEmpty();
+
+        while (childPagesIterator.hasNext()) {
+            com.day.cq.wcm.api.Page childPage = childPagesIterator.next();
+            boolean found = noPageFilters;
+
+            for (Pattern pageFilterPattern : structurePatterns) {
+                if (pageFilterPattern.matcher(childPage.getPath()).find()) {
+                    found = true;
+                    break;
+                }
+            }
+
+            if (!found) {
+                continue;
+            }
+
+            pages.add(childPage);
+
+            pages.addAll(getChildPageRecursive(childPage, slingRequest, structurePatterns, depth));
+        }
+
+        return pages;
+    }
+
+    /**
+     * Optionally add a child page that is the entry point of a site model request when this child is not added by
```

Comment:
From reading the javadocs and the code, it seems a bit magical which page is added to the list of pages.
Could you please clarify the javadocs, maybe add an example?